aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -24,12 +24,39 @@ using namespace at::sparse;
 
 namespace {
 
-  inline SparseTensor get_result_tensor_for_unary_op(const SparseTensor& input) {
-    if (c10::isIntegralType(input.scalar_type(), /*includeBool=*/true)) {
+  template <bool promotes_integer_to_float = false>
+  inline SparseTensor get_result_tensor_for_unary_op(const SparseTensor& input){
+    if (c10::isIntegralType(input.scalar_type(), /*includeBool=*/true) && promotes_integer_to_float) {
       return at::empty_like(input, input.options().dtype(c10::get_default_dtype()));
     }
     return at::empty_like(input);
   }
+
+  // --------------------------------------------------------------------
+  // helper function for sanity check of input and output tensors
+  // this functions also inits the output tensor
+  // this helper function is for unary ops
+  // --------------------------------------------------------------------
+  template <bool promotes_integer_to_float = false, bool is_linear_map = false>
+  inline void sparse_unary_op_check_and_init(const SparseTensor& self, SparseTensor& result){
+    TORCH_CHECK(result.is_sparse(), "Tensor should be sparse");
```

Comment:
Let's tweak these error messages slightly:

"Expected result tensor to be sparse"
"Expected input tensor to be sparse"